File: server/server/src/main/java/model/Personage.java
package model;

import java.io.Serializable;
import java.util.Date;

public abstract class Personage implements Comparable<Personage>, Serializable{
    public String name = "NoName";
    public String type = "Персонаж";
    public double x = 0;
    public double y = 0;
    public int skillSwear;
    public int force;
    public int height;
    public Mood mood = Mood.NORMAL;
    public Date dateCreate;

    @Override
    public boolean equals(Object s) {
        if (s == null) {
            return false;
        }
        if (this == s) {
            return true;
        }
        if (!(getClass() == s.getClass())){
            return false;
        }
        else {
            Personage tmp = (Personage) s;
            return (tmp.name.equals(name) && force == tmp.force && mood == tmp.mood && tmp.type.equals(type));
        }
    }

    @Override
    public int hashCode(){
        final int power = 239;
        int hash = 0;
        for (int i = 0; i < name.length(); ++i) {
            hash = hash * power + (int)name.charAt(i);
        }

        for (int i = 0; i < type.length(); ++i) {
            hash = hash * power + (int)type.charAt(i);
        }


        if (mood == Mood.NORMAL) {
            hash *= 2;
        } else if (mood == Mood.FURY) {
            hash *= 3;
        } else if (mood == Mood.HAPPY) {
            hash *= 4;
        } else if (mood == Mood.SAD) {
            hash *= 5;
        }

        hash *= force * x * y * height;
        return hash;
    }

    @Override
    public String toString(){
        return type + " " + name ;
    }

    /**
     * Кто-то что-то почувствовал
     */
    public void feel(){
        System.out.print(name + " почувствовал, что ");
    }

    /**
     * Метод, сравнивает 2-х персонажей
     * @param p - персонаж с которым сравниваем
     * @return 0 - объекты равны, > 0 => (this > p), < 0 => (this < p)
     */
    @Override
    public int compareTo (Personage p){
        if (this.toString().compareTo(p.toString()) == 0){
            return this.height - p.height;
        } else {
            return this.toString().compareTo(p.toString());
        }
    }

    public boolean setMood(String s) {
        switch (s){
            case "Mood.NORMAL" : {}
            case "NORMAL" : {
                this.mood = Mood.NORMAL;
                break;
            }
            case "Mood.FURY" : {}
            case "FURY" : {
                this.mood = Mood.FURY;
                break;
            }
            case "Mood.HAPPY" : {}
            case "HAPPY" : {
                this.mood = Mood.HAPPY;
                break;
            }
            case "Mood.SAD" : {}
            case "SAD" : {
                this.mood = Mood.SAD;
                break;
            }
            default: {
                return false;
            }
        }
        return true;
    }
}

File: server/server/src/main/kotlin/model/Personage.kt
package model

import orm.*
import orm.Enum
import java.io.Serializable
import java.time.LocalDateTime


@Table(name = "Personages")
open class Personage(
        var type: String = "Персонаж",
        var name: String = "NoName",
        var x: Double = 0.0,
        var y: Double = 0.0,
        var force: Int,
        var height: Int,
        var skillSwear: Int = 0,
        @Enum var mood: Mood,
        var dateCreate: LocalDateTime) : Comparable<Personage>, Serializable {

    override fun equals(s: Any?): Boolean {
        if (s == null) {
            return false
        }
        if (this === s) {
            return true
        }
        if (javaClass != s.javaClass) {
            return false
        } else {
            val tmp = s as Personage?
            return tmp!!.name == name && force == tmp.force && mood == tmp.mood && tmp.type == type
        }
    }

    override fun hashCode(): Int {
        val power = 239
        var hash = 0
        for (i in 0 until name.length) {
            hash = hash * power + name[i].toInt()
        }

        for (i in 0 until type.length) {
            hash = hash * power + type[i].toInt()
        }


        if (mood == Mood.NORMAL) {
            hash *= 2
        } else if (mood == Mood.FURY) {
            hash *= 3
        } else if (mood == Mood.HAPPY) {
            hash *= 4
        } else if (mood == Mood.SAD) {
            hash *= 5
        }

        hash *= (force.toDouble() * x * y * height.toDouble()).toInt()
        return hash
    }

    override fun toString(): String {
        return "$type $name"
    }

    /**
     * Кто-то что-то почувствовал
     */
    fun feel() {
        print("$name почувствовал, что ")
    }

    /**
     * Метод, сравнивает 2-х персонажей
     * @param p - персонаж с которым сравниваем
     * @return 0 - объекты равны, > 0 => (this > p), < 0 => (this < p)
     */
    override fun compareTo(p: Personage): Int {
        return if (this.toString().compareTo(p.toString()) == 0) {
            this.height - p.height
        } else {
            this.toString().compareTo(p.toString())
        }
    }

    fun setMood(s: String): Boolean {
        when (s) {
            "Mood.NORMAL" -> {
                run { }
                run {
                    this.mood = Mood.NORMAL
                }
            }
            "NORMAL" -> {
                this.mood = Mood.NORMAL
            }
            "Mood.FURY" -> {
                run { }
                run {
                    this.mood = Mood.FURY
                }
            }
            "FURY" -> {
                this.mood = Mood.FURY
            }
            "Mood.HAPPY" -> {
                run { }
                run {
                    this.mood = Mood.HAPPY
                }
            }
            "HAPPY" -> {
                this.mood = Mood.HAPPY
            }
            "Mood.SAD" -> {
                run { }
                run {
                    this.mood = Mood.SAD
                }
            }
            "SAD" -> {
                this.mood = Mood.SAD
            }
            else -> {
                return false
            }
        }
        return true
    }
}
